src/_pytest/terminal.py

```diff
@@ -1012,6 +1030,54 @@ def show_skipped(lines: List[str]) -> None:
             for line in lines:
                 self.write_line(line)
 
+    def _get_main_color(self) -> Tuple[str, List[str]]:
+        if self._main_color is None or self._known_types is None or self._is_last_item:
+            self._set_main_color()
+            assert self._main_color
+            assert self._known_types
+        return self._main_color, self._known_types
+
+    def _determine_main_color(self, unknown_type_seen: bool) -> str:
+        stats = self.stats
+        if "failed" in stats or "error" in stats:
+            main_color = "red"
+        elif "warnings" in stats or "xpassed" in stats or unknown_type_seen:
+            main_color = "yellow"
+        elif "passed" in stats or not self._is_last_item:
+            main_color = "green"
+        else:
+            main_color = "yellow"
+        return main_color
+
+    def _set_main_color(self) -> Tuple[str, List[str]]:
+        unknown_types = []  # type: List[str]
+        for found_type in self.stats.keys():
+            if found_type:  # setup/teardown reports have an empty key, ignore them
+                if found_type not in KNOWN_TYPES and found_type not in unknown_types:
+                    unknown_types.append(found_type)
+        self._known_types = list(KNOWN_TYPES) + unknown_types
+        self._main_color = self._determine_main_color(bool(unknown_types))
+        return self._main_color, self._known_types
+
+    def build_summary_stats_line(self) -> Tuple[List[Tuple[str, Dict[str, bool]]], str]:
+        main_color, known_types = self._get_main_color()
+
+        parts = []
+        for key in known_types:
```

Comment:
Following my comment about `known_types` might not being needed, can we use this instead?

```python
known_types = list(KNOWN_TYPES) + [x for x in self.stats if x not in KNOWN_TYPES]
```

Reply:
Could be done maybe.. but a) order is important here (could be solved), but b) we have it already, so can use it.
This can be done in a next round of cleanup/refactoring then maybe.
Also keep in mind that e.g. pytest-parallel needs this to be usable via a multiprocessing proxy.
I'd rather leave it as is for now, but thanks for the suggestion and review, of course.

Reply:
Fair enough. 👍